homeassistant/components/stream/worker.py

```diff
@@ -11,104 +11,190 @@
+            )
 
     def set_streams(self, video_stream, audio_stream):
         """Initialize output buffer with streams from container."""
-        self._video_stream = video_stream
-        self._audio_stream = audio_stream
+        self._input_video_stream = video_stream
+        self._input_audio_stream = audio_stream
 
     def reset(self, video_pts):
         """Initialize a new stream segment."""
         # Keep track of the number of segments we've processed
         self._sequence += 1
-        self._segment_start_pts = video_pts
-
-        # Fetch the latest StreamOutputs, which may have changed since the
-        # worker started.
-        self._outputs = self._outputs_callback().values()
-        self._stream_buffer = create_stream_buffer(
-            self._video_stream, self._audio_stream, self._sequence
-        )
+        self._segment_start_pts = (
+            self._part_start_dts
+        ) = self._last_packet_dts = video_pts
+        self._part_is_independent = False
+        self._last_packet_independent = False
+        self._segment = Segment(sequence=self._sequence, stream_id=self._stream_id)
+        self.reset_av()
 
     def mux_packet(self, packet):
         """Mux a packet to the appropriate StreamBuffers."""
 
         # Check for end of segment
-        if packet.stream == self._video_stream and packet.is_keyframe:
-            duration = (packet.pts - self._segment_start_pts) * packet.time_base
-            if duration >= MIN_SEGMENT_DURATION:
-                # Save segment to outputs
-                self.flush(duration)
+        if packet.stream == self._input_video_stream:
+            segment_duration = (packet.pts - self._segment_start_pts) * packet.time_base
+
+            self.check_flush_part(packet)
+
+            if packet.is_keyframe and segment_duration >= self._min_segment_duration:
+                # Flush segment (also flushes the stub part segment)
+                self.flush(segment_duration, packet)
 
                 # Reinitialize
                 self.reset(packet.pts)
 
+            self._last_packet_dts = packet.dts
+            self._part_is_independent |= self._last_packet_independent
+            self._last_packet_independent = packet.is_keyframe
+
         # Mux the packet
-        if packet.stream == self._video_stream:
-            packet.stream = self._stream_buffer.vstream
-            self._stream_buffer.output.mux(packet)
-        elif packet.stream == self._audio_stream:
-            packet.stream = self._stream_buffer.astream
-            self._stream_buffer.output.mux(packet)
-
-    def flush(self, duration):
+        if packet.stream == self._input_video_stream:
+            packet.stream = self._output_video_stream
+            self._av_output.mux(packet)
+        elif packet.stream == self._input_audio_stream:
+            packet.stream = self._output_audio_stream
+            self._av_output.mux(packet)
+
+    def check_flush_part(self, packet):
```

Comment:
This is new. We need to check if the part is done yet. We can do this by checking whether the position in `self._memory_file` has changed. If so then we can go ahead with the mechanics of copying out and putting the `Part` and resetting the part attributes.